README.txt
flooding-base
==========================================

Introduction

Usage, etc.

More details in src/flooding_base/USAGE.txt .


Requirements
------------

PIL needs to be installed.  (On ubuntu: "apt-get install python-imaging").

Matplotlib also ("python-matplotlib").


Development installation
------------------------

The first time, you'll have to run the "bootstrap" script to set up setuptools
and buildout::

    $> python bootstrap.py

And then run buildout to set everything up::

    $> bin/buildout

(On windows it is called ``bin\buildout.exe``).

You'll have to re-run buildout when you or someone else made a change in
``setup.py`` or ``buildout.cfg``.

The current package is installed as a "development package", so
changes in .py files are automatically available (just like with ``python
setup.py develop``).

If you want to use trunk checkouts of other packages (instead of released
versions), add them as an "svn external" in the ``local_checkouts/`` directory
and add them to the ``develop =`` list in buildout.cfg.

Tests can always be run with ``bin/test`` or ``bin\test.exe``.


CHANGES.txt
Changelog of flooding-base
==========================


1.59 (unreleased)
-----------------

- Nothing changed yet.


1.58 (2014-04-23)
-----------------

- Edit measures on double klick by new scenario

- Select, deselect measures on klick by new scenario

- Fix visibility of 'Aangepaste Keringen' layer


1.57 (2014-04-14)
-----------------

- Nothing changed yet.


1.56 (2014-04-09)
-----------------

- Nothing changed yet.


1.55 (2014-04-09)
-----------------

- Fix overlayes visibility for Chrome and Firefox.


1.54 (2014-04-04)
-----------------

- Add translated stings.
- Extend width of the animation slider.
- Set OpenStreetMap as default baselayer.

1.53 (2014-02-14)
-----------------

- Add translation.


1.52 (2014-02-03)
-----------------

- Fix shimmering effect in animation for IE.


1.51 (2014-01-21)
-----------------

- Fix pyramid's tooltip in IE.


1.50 (2013-12-03)
-----------------

- Add pyramids.


1.49 (2013-11-29)
-----------------

- Fix bug on resize infowindow.
- Fix width of infowindow tabs.


1.48 (2013-11-20)
-----------------

- Select breach in navigation tree on click.


1.47 (2013-11-18)
-----------------

- Add search functionality.a


1.46 (2013-10-23)
-----------------

- Fix urls for table app.


1.45 (2013-10-21)
-----------------

- Add 'Archief' tab into infowindow.


1.44 (2013-10-07)
-----------------

- Change workspace for pdok maps.


1.43 (2013-09-24)
-----------------

- Fix styles.


1.42 (2013-09-23)
-----------------

- Fix updateSize of map on resize window.

- Fix layout of LayerSwitcher.

- Update OpenLayers to 2.12

- Add top10nl as baselayer.

- Handle error due TRANSPORT_ERROR by bad wifi connection in startup.js.


1.41 (2013-09-10)
-----------------

- Change layers field to TextFiled type.


1.40 (2013-09-09)
-----------------

- Update smartclient to v9.0.

- Add speed control button for animations.


1.39 (2013-08-14)
-----------------

- Set projection to maplayer in config.js.


1.38 (2013-07-12)
-----------------

- Fix OpenStreetMap layer.


1.37 (2013-06-18)
-----------------

- Fix bug in NGeneral.js.


1.36 (2013-06-12)
-----------------

- Extended intervalFormatter to use an interval > 30 days.

1.35 (2012-12-20)
-----------------

- Fix widht of IFrame for Table, Importeren, Exporteren (removed float:left style from flooding_base/static/weblib/isomorphic/system/modules/ISC_Core.js


1.34 (2012-12-06)
-----------------

- Added functionality to use a 3di form on creating 3di scenario.

1.33 (2012-11-20)
-----------------

Added gaug.es tracking Javascript to the gui index.html because Joep
wants to track the site.


1.32 (2012-10-19)
-----------------

Add Javascript that can open a specific project/scenario upon loading
the homepage.

Get the information about which scenario to open from the session (it
is set in a view in flooding-lib).

Some cleanup in Javascript.


1.31 (2012-09-21)
-----------------

Changed Javascript, now showing an (I) icon link after some breaches
(only works with new services.py from flooding-lib).


1.30 (2012-08-16)
-----------------

Increased length of slug field of Text model, 20 was too short.


1.29 (2012-07-31)
-----------------

Fixed bug with showing result animations; didn't change rootUrl to
flooding_config.root_url in some places.


1.28 (2012-07-26)
-----------------

Refactor Text.get() so that it works nicely with templates (it always
returns SafeString instances so they aren't escaped anymore in the
template) and gives a helpful reminder text where they are used in the
template while there is no text entered in the admin yet.


1.27 (2012-07-19)
-----------------

Cleaned up index.html. Made header background image work and uses the
current logo now. Moved most of the Javascript that used to be in
index.html into separate files.

Removed a lot of the Application/Subapplication code as this isn't
Lizard anymore, it is now purely Flooding and it doesn't need to be as
dynamic anymore.

Removed much of the code for FEWS, NHI, etc.


1.26 (2012-07-09)
-----------------

- Removed static/examples files, moved them to the flooding site. This
  makes it slightly easier to update them.


1.25 (2012-06-14)
-----------------

- New versions of the group import example Excel files.


1.24 (2012-06-12)
-----------------

- Add 'Text' model to store bits of text that need to be shown on the site.


1.23 (2012-06-12)
-----------------

- Nothing changed yet.


1.22.3 (2012-06-08)
-------------------

- Nothing changed yet.


1.22.2 (2012-06-08)
-------------------

- Nothing changed yet.


1.22.1 (2012-06-08)
-------------------

- Nothing changed yet.


1.22 (2012-06-08)
-----------------

- MOVED TO GIT

- RENAMED FROM lizard-base TO flooding-base


1.21 (2012-05-16)
-----------------

- Nothing changed yet.


1.20 (2012-05-16)
-----------------

- Nothing changed yet.


1.19 (2012-05-10)
-----------------

- New user manual.

- "Export" button disable/enable when a scenario is selected/deselected.

- Typo fix.


1.18 (2012-05-03)
-----------------

- Nothing changed yet.


1.17 (2012-03-06)
-----------------

- Nothing changed yet.


1.16 (2012-03-01)
-----------------

- Nothing changed yet.


1.15 (2012-03-01)
-----------------

- Nothing changed yet.


1.14 (2011-05-25)
-----------------

- Nothing changed yet.


1.13 (2011-05-06)
-----------------

- Nothing changed yet.


1.12 (2011-04-18)
-----------------

- Nothing changed yet.


1.11 (2011-03-25)
-----------------

- Nothing changed yet.


1.10 (2011-02-07)
-----------------

- Making ready for releasing on Lizard-PyPi


1.9 (2011-02-07)
----------------

- Nothing changed yet.


1.8 (2011-02-01)
----------------

- Nothing changed yet.


1.7 (2011-01-20)
----------------

- Nothing changed yet.


1.6 (2010-03-22)
----------------

- Fix service_get_locations.

1.5 (2010-03-19)
----------------

- Using RequestContext in several render_to_response() calls to gain the

flooding_base/models.py
# -*- coding: utf-8 -*-
from django.conf import settings
from django.db import models
from django.contrib.auth.models import Group
from django.utils.translation import ugettext_lazy as _
from django.utils.html import escape
from django.utils.safestring import SafeString

from flooding_base.eidatabaseconnector import ConnectDatabase2EiServer
from flooding_base.dummydatabaseconnector import DummyDatabaseConnector


class Configuration(models.Model):
    """Stores the datasource type and some meta data"""

    DATASOURCE_TYPE_CHOICES = (
        (1, _('DataSourceEI')),
        (2, _('DataSourceDummy')),
        )
    DATASOURCE_TYPE_EI = 1
    DATASOURCE_TYPE_DUMMY = 2

    name = models.CharField(max_length=200)
    description = models.TextField(blank=True)

    # datasource_type tells us to link in the DataSourceEI table or a
    # different table
    datasourcetype = models.IntegerField(choices=DATASOURCE_TYPE_CHOICES)

    groups = models.ManyToManyField(Group,
                                    through='GroupConfigurationPermission')

    #geographic coordinates
    coords_w = models.FloatField(default=3.289)
    coords_e = models.FloatField(default=7.314)
    coords_s = models.FloatField(default=50.765)
    coords_n = models.FloatField(default=53.471)

    class Meta:
        verbose_name = _('Configuration')
        verbose_name_plural = _('Configurations')
        db_table = 'base_configuration'

    def __unicode__(self):
        return self.name

    def getConnector(self):
        """Return connector, according to datasourcetype"""
        if self.datasourcetype == self.DATASOURCE_TYPE_EI:
            return self.datasourceei.getConnector()
        elif self.datasourcetype == self.DATASOURCE_TYPE_DUMMY:
            return self.datasourcedummy.getConnector()
        return None

    def get_datasource(self):
        """Return datasource object"""
        if self.datasourcetype == self.DATASOURCE_TYPE_EI:
            return self.datasourceei
        elif self.datasourcetype == self.DATASOURCE_TYPE_DUMMY:
            return self.datasourcedummy
        return None

    def hasDataSource(self):
        try:
            self.getConnector()
            return True
        except:
            return False

    def getSpecificData(self):
        """Get datasource specific data in a dictionary"""
        if self.datasourcetype == self.DATASOURCE_TYPE_EI:
            return self.datasourceei.getSpecificData()
        elif self.datasourcetype == self.DATASOURCE_TYPE_DUMMY:
            return self.datasourcedummy.getSpecificData()
        return None


class DataSourceEI(models.Model):
    """datasource types for EI connection (i.e. Jdbc2Ei)"""

    configuration = models.OneToOneField(Configuration, unique=True)
    # configuration must be unique
    connectorurl = models.CharField(
        max_length=200,
        default='http://127.0.0.1:8080/Jdbc2Ei/test')
    databaseurl = models.CharField(
        max_length=200,
        default='jdbc:vjdbc:rmi://127.0.0.1:2000/VJdbc,FewsDataStore')
    databaseurltagname = models.CharField(max_length=200)

    usecustomfilterresponse = models.BooleanField(default=False)
    customfilterresponse = models.TextField(
        null=True, blank=True,
        help_text=("Use a pythonic list of dictionaries. The rootnode has "
                   "'parentid': None. i.e. [{'id':'id','name':'name',"
                   "'parentid':None}, {'id':'id2','name':'name2',"
                   "'parentid':'id'}]"))

    class Meta:
        verbose_name = _('DataSourceEI')
        verbose_name_plural = _('Data sources EI')
        db_table = 'base_datasourceei'

    def __unicode__(self):
        return self.configuration.__unicode__()

    def getConnector(self):
        return ConnectDatabase2EiServer(self.connectorurl,
                                        self.databaseurl,
                                        self.databaseurltagname)

    def getSpecificData(self):
        connector = self.getConnector()
        try:
            databaseurl_actual = connector.getUrl()
        except:
            databaseurl_actual = 'Could not retrieve actual databaseurl'
        return {'databaseurl': self.databaseurl,
                'databaseurltagname': self.databaseurltagname,
                'databaseurl_actual': databaseurl_actual}


class DataSourceDummy(models.Model):
    """datasource type dummy connection"""

    configuration = models.OneToOneField(Configuration, unique=True)
    # configuration must be unique

    class Meta:
        verbose_name = _('DataSourceDummy')
        verbose_name_plural = _('Data sources dummy')
        db_table = 'base_datasourcedummy'

    def __unicode__(self):
        return self.configuration.__unicode__()

    def getConnector(self):
        return DummyDatabaseConnector()

    def getSpecificData(self):
        return {}


class Application(models.Model):
    """Lists all available applications"""
    TYPE_FEWS = 1
    TYPE_FLOODING = 2
    TYPE_FLOW = 3
    TYPE_PRESENTATION = 4
    TYPE_GISVIEWER = 5
    TYPE_MIS = 6
    TYPE_NHI = 7

    TYPE_CHOICES = (
        (TYPE_FEWS, 'fews'),
        (TYPE_FLOODING, 'flooding'),
        (TYPE_FLOW, 'flow'),
        (TYPE_PRESENTATION, 'presentation'),
        (TYPE_GISVIEWER, 'gisviewer'),
        (TYPE_MIS, 'mis'),
        (TYPE_NHI, 'nhi'),
        )
    TYPE_DICT = dict(TYPE_CHOICES)

    name = models.CharField(max_length=200)
    description = models.TextField(null=True, blank=True)
    type = models.IntegerField(choices=TYPE_CHOICES)
    active = models.BooleanField(default=True)

    class Meta:
        db_table = 'base_application'

    def __unicode__(self):
        return u'%s' % self.name

    def get_application_name(self):
        return self.TYPE_DICT[self.type]

    def get_subapplication_jsnames_str(self):
        """return jsnames in a string, seperated by comma.

        i.e. "fews_map, fews_graphs"

        """
        return ', '.join([sa.get_subapplication_jsname() for sa in
                          self.subapplication_set.all()])

    @classmethod
    def get_applications(cls, site=None):
        """Get all available applications. Optionally gives the
        applications for a given site.

        Returns a dictionary of key: js app name, value: application
        object, where active=True
        """
        if site is None:
            objects = cls.objects.filter(active=True)
        else:
            objects = Application.objects.filter(
                subapplication__in=site.subapplications.all(),
                active=True).distinct()
        return dict([(a.get_application_name(), a) for a in objects])


class SubApplication(models.Model):
    """
    Subapplications of application. Used by Site as starter application.

    DO NOT CHANGE TYPE_CHOICES WITHOUT KNOWING THE CONSEQUENCES
    """

    TYPE_CHOICES = (
        (11, 'fewsMap'),
        (12, 'fewsGraph'),
        (13, 'fewsReport'),
        (14, 'fewsConfig'),
        (21, 'floodingResults'),
        (22, 'floodingNew'),
        (23, 'floodingTable'),
        (24, 'floodingImport'),
        (25, 'floodingExport'),
        (31, 'flowResults'),
        (41, 'gisviewerMap'),
        (51, 'mis_map'),
        (52, 'mis_table'),
        (53, 'mis_report'),
        (54, 'mis_admin'),
        (55, 'mis_import'),
        (61, 'nhi_map'),
        )
    TYPE_DICT = dict(TYPE_CHOICES)
    NAME2TYPE = dict([(b, a) for (a, b) in TYPE_CHOICES])
    JSNAMES = {
        11: 'fews_map',
        12: 'fews_graphs',
        13: 'fews_report',
        14: 'fews_lcm',
        21: 'flooding_result',
        22: 'flooding_new',
        23: 'flooding_table',
        24: 'flooding_import',
        25: 'flooding_export',
        31: 'flow_result',
        41: 'gisviewer_map',
        51: 'mis_map',
        52: 'mis_table',
        53: 'mis_report',
        54: 'mis_admin',
        55: 'mis_import',
        61: 'nhi_map',
        }

    application = models.ForeignKey(Application)
    index = models.IntegerField(default=1)
    type = models.IntegerField(choices=TYPE_CHOICES)

    class Meta:
        ordering = ('application', 'index')
        db_table = 'base_subapplication'

    def __unicode__(self):
        return u'%s' % self.TYPE_DICT[self.type]

    def get_subapplication_name(self):
        return self.TYPE_DICT[self.type]

    def get_subapplication_jsname(self):
        return self.JSNAMES[self.type]


class GroupConfigurationPermission(models.Model):
    """Permission of groups on configurations"""
    PERMISSION_CHOICES = (
        (1, _('VIEW')),
        )

    PERMISSION_VIEW = 1

    configuration = models.ForeignKey(Configuration)
    group = models.ForeignKey(Group)
    permission = models.IntegerField(choices=PERMISSION_CHOICES, default=1)

    class Meta:
        db_table = 'base_groupconfigurationpermission'

    def __unicode__(self):
        return u'%s %s %s' % (self.group, self.configuration, self.permission)


class Setting(models.Model):
    """Stores settings for the project"""
    key = models.CharField(max_length=200, unique=True)
    value = models.CharField(max_length=200)
    remarks = models.TextField(null=True, blank=True)

    class Meta:
        db_table = 'base_setting'

    def __unicode__(self):
        return u'%s = %s' % (self.key, self.value)


class Map(models.Model):
    """Stores wms entries"""
    name = models.CharField(max_length=200)
    remarks = models.TextField(blank=True, null=True)
    active = models.BooleanField(default=True)

    url = models.CharField(max_length=200)

    #layernames separated with comma
    layers = models.CharField(max_length=200)

    transparent = models.NullBooleanField(default=None)
    is_base_layer = models.NullBooleanField(default=False)
    tiled = models.NullBooleanField(default=None)
    srs = models.CharField(max_length=50, default='EPSG:900913')

    class Meta:
        db_table = 'base_map'

    def __unicode__(self):
        return self.name

#possible extension: PermissionGroupSubApplication


class Site(models.Model):
    """Pre-configured sites"""
    name = models.CharField(max_length=200)
    configurations = models.ManyToManyField(Configuration,
                                            blank=True,
                                            null=True)

    maps = models.ManyToManyField(Map, null=True, blank=True)
    subapplications = models.ManyToManyField(SubApplication)
    #starter_app: subapp, because the app is then automatically defined too
    starter_application = models.ForeignKey(
        SubApplication,
        related_name='starter_application')

    favicon_image = models.ImageField(
        upload_to='user/favicon', null=True, blank=True)
    logo_image = models.ImageField(
        upload_to='user/logo', null=True, blank=True,
        help_text=_('The size is 280x45 pixels'))
    topbar_image = models.ImageField(
        upload_to='user/topbar', null=True, blank=True,
        help_text=_('The size is 600x70 pixels (width can vary)'))

    coords_w = models.FloatField(default=3.289)
    coords_e = models.FloatField(default=7.314)
    coords_s = models.FloatField(default=50.765)
    coords_n = models.FloatField(default=53.471)

    class Meta:
        db_table = 'base_site'

    def __unicode__(self):
        return self.name

    def get_applications(self):
        return Application.get_applications(site=self)

    def get_subapplication_jsnames_str(self):
        """return dict with per application the list of
        subapplications in a single string"""
        result = {}
        for app_name, a in self.get_applications().items():
            result[app_name] = ', '.join(
                [sa.get_subapplication_jsname()
                 for sa in self.subapplications.filter(application=a)])
        return result


class Text(models.Model):
    """Some text to show on the site, somewhere."""

    LANGUAGE_CHOICES = (('nl', 'Nederlands'),)
    slug = models.CharField(max_length=200)
    language = models.CharField(
        max_length=5, default="nl",
        choices=LANGUAGE_CHOICES)
    content = models.TextField(blank=True)
    is_html = models.BooleanField(default=False)

    @classmethod
    def get(cls, slug, language_code=None, request=None):
        """Returns the text's content, HTML-escaped unless it is
        itself HTML, and marked as a SafeString so that it can be used
        in templates immediately.

        If language_code is not given, it can be retrieved from the
        request, if that is given.

        If the text isn't found, a short text saying which slug should be
        entered into the admin interface is returned."""

        if language_code is None and request:
            language_code = getattr(request, 'LANGUAGE_CODE', None)

        search = {
            'slug': slug,
        }
        if language_code is not None:
            search['language'] = language_code

        texts = cls.objects.filter(**search)
        if texts.count() > 0:
            text = texts[0]
            if text.is_html:
                return SafeString(text.content)
            else:
                return SafeString(escape(text.content))
        else:
            return SafeString(escape(
        "Text not found. Enter a Text with slug '{0}' and language_code '{1}'."
                .format(slug, language_code if language_code else '')))

    def __unicode__(self):
        return "{slug} ({language})".format(
            slug=self.slug, language=self.language)
